Fix sign(NAN) returning 1.

AcatXIo opened the pull request:
Fixes godotengine#79036 by using the same implementation for SIGN() as numpy does.
Adds a test for the NAN case. Updates the documentation to clarify the new behavior.

AcatXIo commented:
I tested the performance of this in a compiled Windows editor with the following code:
```gdscript
func _ready():
	var _a
	var start_time
	var end_time
	await get_tree().create_timer(1).timeout
	start_time = Time.get_ticks_usec()
	for y in range(100000000):
		for x in range(40):
			_a = sign(32.0)
		for x in range(39):
			_a = sign(-15.0)
		for x in range(20):
			_a = sign(0.0)
		for x in range(1):
			_a = sign(NAN)
	end_time = Time.get_ticks_usec()
	print("Running time: " + str(end_time - start_time) + "μs.")
```
The resulting run time was 637.82s for 4.2.dev2 and 597.97s for this PR.

AThousandShips commented on tests/core/math/test_math_funcs.h:
```diff
@@ -54,6 +54,8 @@ TEST_CASE("[Math] C++ macros") {
 	CHECK(SIGN(0) == 0.0);
 	CHECK(SIGN(5) == 1.0);
+	// check that SIGN(NAN) returns NAN.
+	CHECK(SIGN(NAN) != SIGN(NAN));
```
```suggestion
	CHECK(Math::is_nan(SIGN(NAN)));
```
This doesn't check anything, as any return value will pass this check, you need to use explicit checking

AcatXIo replied:
Fixed.

akien-mga commented:
> The resulting run time was 637.82s for 4.2.dev4 and 597.97s for this PR.

Are you using official builds for 4.2.dev4, or self-compiled? A benchmark only makes sense if the two binaries are compiled with the exact same toolchains.

Commit:
Fix sign(NAN) returning 1.

Fixes godotengine#79036 by using the same implementation for SIGN() as numpy does.
This can also improve performance, assuming that the most frequently passed values are positive, followed by negative, then zero, and the rarest ones are NAN.
Adds a test for the NAN case. Updates the documentation to clarify the new behavior.

AcatXIo commented:
I was using the official one, I will redo the test on a self compiled version now and send the results.

AcatXIo commented:
Sorry for the confusion, but the PR actually decreases performance quite significantly.
The resulting running times were 549.97 seconds for current master, and 597.97 for this PR (both custom-built).

If there is anything I can do to improve this, please let me know, otherwise I'm going to close the PR, and try to write an alternative solution.

AThousandShips commented:
We also need to investigate if this breaks any code

There could be code paths that depend on the value of `sign` that would break if the return value was `NaN`, I think it's safer to just expect the user to do `NaN` checks if they need to, possibly changing only the behaviour in GDScript

But I think the performance drop, and the potential risks, speak against this change

AcatXIo commented:
Ok, thank you for your time.

AThousandShips commented:
That's just my opinion though, no need to close unless you agree with these concerns

AcatXIo commented:
Ok, I will do some more tests, and possibly move the change to the globalscope function, if it will  work better I will open a new PR.
